placement/placement.go

```diff
@@ -0,0 +1,667 @@
+		if db == nil {
+			return errDatabaseNotFound
+		}
+
+		// If this is an existing cluster, add to the list of decomissions if it's not already there
+		if _, existing := db.Clusters[cName]; existing {
+			if dbChanges.Decomms == nil {
+				dbChanges.Decomms = make(map[string]*schema.ClusterDecommission)
+			}
+
+			sp.log.Infof("decommissioning cluster %s in database %s", cName, dbName)
+			dbChanges.Decomms[cName] = &schema.ClusterDecommission{
+				ClusterName: cName,
+			}
+			return nil
+		}
+
+		// If this is a pending join, delete from the join list
+		if _, joining := dbChanges.Joins[cName]; joining {
+			sp.log.Infof("removing pending cluster %s from database %s", cName, dbName)
+			delete(dbChanges.Joins, cName)
+			return nil
+		}
+		return errClusterNotFound
+	}))
+}
+
+func (sp storagePlacement) GetPendingChanges() (int, *schema.Placement, *schema.PlacementChanges, error) {
+	vers, protoConfig, protoChanges, err := sp.mgr.GetPendingChanges()
+
+	var p *schema.Placement
+	if protoConfig != nil {
+		p = protoConfig.(*schema.Placement)
+	}
+
+	var changes *schema.PlacementChanges
+	if protoChanges != nil {
+		changes = protoChanges.(*schema.PlacementChanges)
+	}
+
+	return vers, p, changes, err
+}
+
+func (sp storagePlacement) CommitChanges(version int, opts CommitOptions) error {
+	return sp.mgr.Commit(version, wrapFn(func(p *schema.Placement, changes *schema.PlacementChanges) error {
+		sp.log.Infof("committing changes at version %d", version)
+
+		// Pull all existing databases, and order them by max retention period.  This will be used
+		// to determine if any new databases are pulling traffic from an existing database
+		existingDatabases := make([]*schema.Database, 0, len(p.Databases))
+		for _, existing := range p.Databases {
+			existingDatabases = append(existingDatabases, existing)
+		}
+		sort.Sort(databasesByRetention(existingDatabases))
+
+		// Add new databases
+		for _, add := range changes.DatabaseAdds {
+			if err := sp.commitNewDatabase(p, add.Database, existingDatabases, opts); err != nil {
+				return err
+			}
+		}
+
+		// Process database changes, including those from new databases
+		for name, change := range changes.DatabaseChanges {
+			if err := sp.commitDatabaseChanges(p, p.Databases[name], change, opts); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}))
+}
+
+func (sp storagePlacement) commitNewDatabase(
+	p *schema.Placement,
+	db *schema.Database,
+	existingDatabases []*schema.Database,
+	opts CommitOptions) error {
+	sp.log.Infof("Creating initial database configuration for %s", db.Name)
+
+	// Check to see if there are any existing databases that cover this retention
+	// period - if so then we need to do a staged cutover to that database
+	for _, existing := range existingDatabases {
+		if existing.MaxRetentionInSecs < db.MaxRetentionInSecs {
+			// We're not pulling traffic from this database
+			continue
+		}
+
+		var (
+			readCutoverTime     = sp.clock.Now().Add(opts.GetRolloutDelay())
```

Comment:
The commit is atomic, but there's no guarantee that readers and writers
will transition at the same time.  I could make this be based on transition
delay to shorten the cutover.

On Sun, Aug 28, 2016 at 12:32 PM, xichen2020 <notifications@github.com>
wrote:

> In placement/placement.go
> <https://github.com/m3db/m3storage/pull/5#discussion_r76535012>:
>
> > +	p *schema.Placement,
> > +	db *schema.Database,
> > +	existingDatabases []*schema.Database,
> > +	opts CommitOptions) error {
> > +	sp.log.Infof("Creating initial database configuration for %s", db.Name)
> > +
> > +	// Check to see if there are any existing databases that cover this retention
> > +	// period - if so then we need to do a staged cutover to that database
> > +	for _, existing := range existingDatabases {
> > +		if existing.MaxRetentionInSecs < db.MaxRetentionInSecs {
> > +			// We're not pulling traffic from this database
> > +			continue
> > +		}
> > +
> > +		var (
> > +			readCutoverTime     = sp.clock.Now().Add(opts.GetRolloutDelay())
>
> Agreed, although in that case why do we need another rollout delay between
> readCutoverTime and writeCutoverTime? Since the commit is atomic, both
> the readers and the writers will receive the new config after roll out
> delay, so seems we should be able to cut over reads and writes at the same
> time?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/m3db/m3storage/pull/5/files/569ecb9b030d3c568e97f1e4f777659c353b3129#r76535012>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAX-WVO2Oh4lJGpI-L-rQla4XybdTN8Uks5qkbgrgaJpZM4Jubit>
> .
>